# Ebay Kleinanzeigen Dataset for Data cleaning Exercise

The dataset was scraped and uploaded to Kaggle. In this exercise, the dataset had been modified as follows:
1. Only 50,000 data points were randomly smapled for the speed.
2. A dirtied version of the dataset to reflect the original status of a scraped dataset.


In [104]:
import numpy as np
import pandas as pd
autos = pd.read_csv("autos.csv", encoding="Latin-1")

In [105]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

## Initial Observation 

After importing to a dataframe: autos, we noticed a few things about this dataset. 
1. There are 20 columns in total, including 5 columns in numeric format and others a mixture of dates, strings and a combination of integers and strings.
2. 15 columns have 50,000 entries while 5 columns have missing entries, null values ranging from (0, 10,000)
3. This dataset takes more than 7.6MB of the memory
4. Some column names contain camelcase instead of using Python's snakecase convention.

Therefore, let's change the column names to better understanding snakecases.

In [ ]:
print(autos.columns)

In [128]:
for c in autos.columns:
    if c == "yearOfRegistration":
        autos.rename({"yearOfRegistration":"registration_year"},inplace=True, axis=1)
    elif c == "monthOfRegistration":
        autos.rename({"monthOfRegistration":"registration_month"},inplace=True, axis=1)
    elif c == "notRepairedDamage":
        autos.rename({"notRepairedDamage":"unrepaired_damage"},inplace=True, axis=1)
    elif c == "dateCreated":
        autos.rename({"dateCreated":"ad_created"},inplace=True, axis=1)
    else:     
        autos.rename({c:c.lower()},inplace=True, axis=1)   

In [129]:
autos.head()

datecrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offertype   price   abtest vehicletype  registration_year  \
0  privat   Angebot  $5,000  control         bus               2004   
1  privat   Angebot  $8,500  control   limousine               1997   
2  privat   Angebot  $8,990     test   limousine               2009   
3  privat   Angebot  $4,350  control  kleinwagen               2007   
4  privat   Angebot  $1,350     test       kombi               2003   

     gearbox  powerps   model   odometer  registration_month fueltype  \
0    manuell      158  andere  150,000km                   3      lpg   
1  automatik      286     7er  150,000km                   6   benzin   
2    manuell      102    golf   70,000km                   7   benzin   
3  automatik       71  fortwo   70,000km                   6   benzin   
4    manuell        0   focus  150,000km                   7   benzin   

        brand unrepaired_damage           ad_created  nrofpictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalcode             lastseen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

### Clean column names

In the above code block, we have changed the camelcased column names to snakecased while modified four of them to shorter and simpler names.

In [130]:
autos.describe(include="all")

datecrawled         name  seller offertype  price abtest  \
count                 50000        50000   50000     50000  50000  50000   
unique                48213        38754       2         2   2357      2   
top     2016-03-09 11:54:38  Ford_Fiesta  privat   Angebot     $0   test   
freq                      3           78   49999     49999   1421  25756   
mean                    NaN          NaN     NaN       NaN    NaN    NaN   
std                     NaN          NaN     NaN       NaN    NaN    NaN   
min                     NaN          NaN     NaN       NaN    NaN    NaN   
25%                     NaN          NaN     NaN       NaN    NaN    NaN   
50%                     NaN          NaN     NaN       NaN    NaN    NaN   
75%                     NaN          NaN     NaN       NaN    NaN    NaN   
max                     NaN          NaN     NaN       NaN    NaN    NaN   

       vehicletype  registration_year  gearbox       powerps  model  \
count        44905       50000.000000    47320  50000.000000  47242   
unique           8                NaN        2           NaN    245   
top      limousine                NaN  manuell           NaN   golf   
freq         12859                NaN    36993           NaN   4024   
mean           NaN        2005.073280      NaN    116.355920    NaN   
std            NaN         105.712813      NaN    209.216627    NaN   
min            NaN        1000.000000      NaN      0.000000    NaN   
25%            NaN        1999.000000      NaN     70.000000    NaN   
50%            NaN        2003.000000      NaN    105.000000    NaN   
75%            NaN        2008.000000      NaN    150.000000    NaN   
max            NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fueltype       brand unrepaired_damage  \
count       50000        50000.000000    45518       50000             40171   
unique         13                 NaN        7          40                 2   
top     150,000km                 NaN   benzin  volkswagen              nein   
freq        32424                 NaN    30107       10687             35232   
mean          NaN            5.723360      NaN         NaN               NaN   
std           NaN            3.711984      NaN         NaN               NaN   
min           NaN            0.000000      NaN         NaN               NaN   
25%           NaN            3.000000      NaN         NaN               NaN   
50%           NaN            6.000000      NaN         NaN               NaN   
75%           NaN            9.000000      NaN         NaN               NaN   
max           NaN           12.000000      NaN         NaN               NaN   

                 ad_created  nrofpictures    postalcode             lastseen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

In [131]:
autos["seller"].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [132]:
autos["offertype"].value_counts()

Angebot    49999
Gesuch         1
Name: offertype, dtype: int64

In [133]:
autos["nrofpictures"].value_counts()

0    50000
Name: nrofpictures, dtype: int64

### Potential Cleaning

After investigating the descriptive information of the dataset, we have detected a few potential cleaning tasks to be completed.
1. column "seller" and "offertype" both have almost one value. Probably not very meaningful for our analysis. These two columns could be dropped. "nrofpictures" has a mono value of 50000 is not very useful and to be dropped.
2. both "price" and "odometer" columns are essentially numbers but disguised in strings. We should turn them into floats for further analysis.
3. "datecrawled" is by seconds which probably are too detailed to be used for our purpose. Potentially, we could turn this into just year-month or year-month-day or just consider them the same and drop this column
4. Obviously there are some mistakes in the column "registration_year". Most likely any number beyond (1900, 2016) is an error and should be marked as "unknown" or "nan". Meanwhile, column "registration_month" has 0s which also requires attention. 
5. Similarly, column "powerps" being 0 is probably an error and need to be adjusted. 

Let's first transfer the strings to integers for "price" and "odometer".

In [134]:
autos["price"] = autos["price"].str.replace("$","").str.replace(",","").astype(float)

In [135]:
autos["odometer"]=autos["odometer"].str.replace("km","").str.replace(",","").astype(int)

In [136]:
autos.rename({"odometer":"odometer_km"},inplace=True, axis=1)

In [137]:
autos.head()

datecrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offertype   price   abtest vehicletype  registration_year  \
0  privat   Angebot  5000.0  control         bus               2004   
1  privat   Angebot  8500.0  control   limousine               1997   
2  privat   Angebot  8990.0     test   limousine               2009   
3  privat   Angebot  4350.0  control  kleinwagen               2007   
4  privat   Angebot  1350.0     test       kombi               2003   

     gearbox  powerps   model  odometer_km  registration_month fueltype  \
0    manuell      158  andere       150000                   3      lpg   
1  automatik      286     7er       150000                   6   benzin   
2    manuell      102    golf        70000                   7   benzin   
3  automatik       71  fortwo        70000                   6   benzin   
4    manuell        0   focus       150000                   7   benzin   

        brand unrepaired_damage           ad_created  nrofpictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalcode             lastseen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [138]:
autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [139]:
autos["price"].value_counts().sort_index().head(50)


0.0      1421
1.0       156
2.0         3
3.0         1
5.0         2
8.0         1
9.0         1
10.0        7
11.0        2
12.0        3
13.0        2
14.0        1
15.0        2
17.0        3
18.0        1
20.0        4
25.0        5
29.0        1
30.0        7
35.0        1
40.0        6
45.0        4
47.0        1
49.0        4
50.0       49
55.0        2
59.0        1
60.0        9
65.0        5
66.0        1
70.0       10
75.0        5
79.0        1
80.0       15
89.0        1
90.0        5
99.0       19
100.0     134
110.0       3
111.0       2
115.0       2
117.0       1
120.0      39
122.0       1
125.0       8
129.0       1
130.0      15
135.0       1
139.0       1
140.0       9
Name: price, dtype: int64

In [140]:
autos.loc[autos["price"]>10000000, :]

datecrawled                                               name  \
2897   2016-03-12 21:50:57   Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000   
24384  2016-03-21 13:57:51                            Schlachte_Golf_3_gt_tdi   
27371  2016-03-09 15:45:47                                         Fiat_Punto   
39377  2016-03-08 23:53:51                        Tausche_volvo_v40_gegen_van   
39705  2016-03-22 14:58:27                        Tausch_gegen_gleichwertiges   
42221  2016-03-08 20:39:05                                  Leasinguebernahme   
47598  2016-03-31 18:56:54  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...   

       seller offertype       price   abtest vehicletype  registration_year  \
2897   privat   Angebot  11111111.0     test   limousine               1973   
24384  privat   Angebot  11111111.0     test         NaN               1995   
27371  privat   Angebot  12345678.0  control         NaN               2017   
39377  privat   Angebot  12345678.0  control         NaN               2018   
39705  privat   Angebot  99999999.0  control   limousine               1999   
42221  privat   Angebot  27322222.0  control   limousine               2014   
47598  privat   Angebot  12345678.0  control   limousine               2001   

         gearbox  powerps     model  odometer_km  registration_month fueltype  \
2897     manuell       48    escort        50000                   3   benzin   
24384        NaN        0       NaN       150000                   0      NaN   
27371        NaN       95     punto       150000                   0      NaN   
39377    manuell       95       v40       150000                   6      NaN   
39705  automatik      224  s_klasse       150000                   9   benzin   
42221    manuell      163        c4        40000                   2   diesel   
47598    manuell      101    vectra       150000                   3   benzin   

               brand unrepaired_damage           ad_created  nrofpictures  \
2897            ford              nein  2016-03-12 00:00:00             0   
24384     volkswagen               NaN  2016-03-21 00:00:00             0   
27371           fiat               NaN  2016-03-09 00:00:00             0   
39377          volvo              nein  2016-03-08 00:00:00             0   
39705  mercedes_benz               NaN  2016-03-22 00:00:00             0   
42221        citroen               NaN  2016-03-08 00:00:00             0   
47598           opel              nein  2016-03-31 00:00:00             0   

       postalcode             lastseen  
2897        94469  2016-03-12 22:45:27  
24384       18519  2016-03-21 14:40:18  
27371       96110  2016-03-09 15:45:47  
39377       14542  2016-04-06 23:17:31  
39705       73525  2016-04-06 05:15:30  
42221       76532  2016-03-08 20:39:05  
47598        4356  2016-03-31 18:56:54

In [141]:
autos.drop(autos[autos["price"] > 10000000].index, inplace=True)

In [142]:
autos.loc[autos["price"]>10000000, :]

Empty DataFrame
Columns: [datecrawled, name, seller, offertype, price, abtest, vehicletype, registration_year, gearbox, powerps, model, odometer_km, registration_month, fueltype, brand, unrepaired_damage, ad_created, nrofpictures, postalcode, lastseen]
Index: []

Drop the two columns "seller" and "offertype"

In [166]:
autos.drop(autos.loc[:,["seller", "offertype"]], inplace = True, axis = 1)

In [168]:
autos.drop(autos.loc[:,["nrofpictures"]], inplace = True, axis = 1)

In [169]:
autos.head()

datecrawled                                               name   price  \
0  2016-03-26                   Peugeot_807_160_NAVTECH_ON_BOARD  5000.0   
1  2016-04-04         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  8500.0   
2  2016-03-26                         Volkswagen_Golf_1.6_United  8990.0   
3  2016-03-12  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...  4350.0   
4  2016-04-01  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...  1350.0   

    abtest vehicletype  registration_year    gearbox  powerps   model  \
0  control         bus               2004    manuell      158  andere   
1  control   limousine               1997  automatik      286     7er   
2     test   limousine               2009    manuell      102    golf   
3  control  kleinwagen               2007  automatik       71  fortwo   
4     test       kombi               2003    manuell        0   focus   

   odometer_km  registration_month fueltype       brand unrepaired_damage  \
0       150000                   3      lpg     peugeot              nein   
1       150000                   6   benzin         bmw              nein   
2        70000                   7   benzin  volkswagen              nein   
3        70000                   6   benzin       smart              nein   
4       150000                   7   benzin        ford              nein   

   ad_created  postalcode    lastseen  
0  2016-03-26       79588  2016-04-06  
1  2016-04-04       71034  2016-04-06  
2  2016-03-26       35394  2016-04-06  
3  2016-03-12       33729  2016-03-15  
4  2016-04-01       39218  2016-04-01

### Dropping rows based on "Price"

Some vehicles have price tags more than 10million and the numbers look artibary, e.g. 12345678. It seems like those are dummy inputs in the system. Therefore, we have decided to delete those 5 records.

Next we turn to the date columns


In [143]:
autos['datecrawled'] = autos['datecrawled'].str[:10]

In [144]:
autos['ad_created'] = autos['ad_created'].str[:10]

In [145]:
autos['lastseen'] = autos['lastseen'].str[:10]

Using the string method above, we have successfully taking the year-month-day for the three date columns for easy analysis.

In [146]:
autos['lastseen'].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001080
2016-03-06    0.004421
2016-03-07    0.005361
2016-03-08    0.007581
2016-03-09    0.009841
2016-03-10    0.010762
2016-03-11    0.012522
2016-03-12    0.023803
2016-03-13    0.008981
2016-03-14    0.012802
2016-03-15    0.015882
2016-03-16    0.016442
2016-03-17    0.027924
2016-03-18    0.007421
2016-03-19    0.015742
2016-03-20    0.020703
2016-03-21    0.020723
2016-03-22    0.021583
2016-03-23    0.018583
2016-03-24    0.019563
2016-03-25    0.019203
2016-03-26    0.016962
2016-03-27    0.016022
2016-03-28    0.020863
2016-03-29    0.022343
2016-03-30    0.024843
2016-03-31    0.023823
2016-04-01    0.023103
2016-04-02    0.024903
2016-04-03    0.025364
2016-04-04    0.024623
2016-04-05    0.124297
2016-04-06    0.220991
2016-04-07    0.130938
Name: lastseen, dtype: float64

"lastseen" column has been created by the crawler in a month time, ranges from 5, March, 2016 to 7, April, 2016.

In [147]:
autos['ad_created'].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.000020
2015-08-10    0.000020
2015-09-09    0.000020
2015-11-10    0.000020
2015-12-05    0.000020
2015-12-30    0.000020
2016-01-03    0.000020
2016-01-07    0.000020
2016-01-10    0.000040
2016-01-13    0.000020
2016-01-14    0.000020
2016-01-16    0.000020
2016-01-22    0.000020
2016-01-27    0.000060
2016-01-29    0.000020
2016-02-01    0.000020
2016-02-02    0.000040
2016-02-05    0.000040
2016-02-07    0.000020
2016-02-08    0.000020
2016-02-09    0.000040
2016-02-11    0.000020
2016-02-12    0.000060
2016-02-14    0.000040
2016-02-16    0.000020
2016-02-17    0.000020
2016-02-18    0.000040
2016-02-19    0.000060
2016-02-20    0.000040
2016-02-21    0.000060
                ...   
2016-03-09    0.033225
2016-03-10    0.031864
2016-03-11    0.032785
2016-03-12    0.036605
2016-03-13    0.016922
2016-03-14    0.035225
2016-03-15    0.033745
2016-03-16    0.030004
2016-03-17    0.031204
2016-03-18    0.013722
2016-03-19    0.033845
2016-03-20    0.037865
2016-03-21 

Most of the ads were created in the period of crawling.

In [148]:
autos['datecrawled'].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025384
2016-03-06    0.013942
2016-03-07    0.035965
2016-03-08    0.033265
2016-03-09    0.033205
2016-03-10    0.032124
2016-03-11    0.032485
2016-03-12    0.036765
2016-03-13    0.015562
2016-03-14    0.036625
2016-03-15    0.033985
2016-03-16    0.029504
2016-03-17    0.031524
2016-03-18    0.013062
2016-03-19    0.034905
2016-03-20    0.037825
2016-03-21    0.037505
2016-03-22    0.032925
2016-03-23    0.032385
2016-03-24    0.029104
2016-03-25    0.031744
2016-03-26    0.032485
2016-03-27    0.031044
2016-03-28    0.034845
2016-03-29    0.034185
2016-03-30    0.033625
2016-03-31    0.031904
2016-04-01    0.033805
2016-04-02    0.035405
2016-04-03    0.038685
2016-04-04    0.036525
2016-04-05    0.013102
2016-04-06    0.003180
2016-04-07    0.001420
Name: datecrawled, dtype: float64

The crawling happened during that same month and rather evenly distributed in that period.

In [149]:
autos['registration_year'].value_counts(normalize=True, dropna=False).sort_index()

1000    0.000020
1001    0.000020
1111    0.000020
1500    0.000020
1800    0.000040
1910    0.000180
1927    0.000020
1929    0.000020
1931    0.000020
1934    0.000040
1937    0.000080
1938    0.000020
1939    0.000020
1941    0.000040
1943    0.000020
1948    0.000020
1950    0.000060
1951    0.000040
1952    0.000020
1953    0.000020
1954    0.000040
1955    0.000040
1956    0.000100
1957    0.000040
1958    0.000080
1959    0.000140
1960    0.000680
1961    0.000120
1962    0.000080
1963    0.000180
          ...   
2001    0.054048
2002    0.050667
2003    0.054548
2004    0.054748
2005    0.060308
2006    0.054168
2007    0.046086
2008    0.044626
2009    0.041966
2010    0.031944
2011    0.032685
2012    0.026464
2013    0.016122
2014    0.013302
2015    0.007981
2016    0.026324
2017    0.029044
2018    0.009821
2019    0.000060
2800    0.000020
4100    0.000020
4500    0.000020
4800    0.000020
5000    0.000080
5911    0.000020
6200    0.000020
8888    0.000020
9000    0.0000

Clearly that some of the years in the list are not valid. Since the dataset was generated in 2016, any number beyong 2016 should be removed. More the car was invented in 1885. Therefore, we can savely remove all records before 1900.

In [150]:
autos.drop(autos[(autos["registration_year"]>2016) | (autos["registration_year"]<1900)].index, inplace=True)

In [151]:
autos['registration_year'].value_counts(normalize=True, dropna=False).sort_index()

1910    0.000187
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000042
1937    0.000083
1938    0.000021
1939    0.000021
1941    0.000042
1943    0.000021
1948    0.000021
1950    0.000062
1951    0.000042
1952    0.000021
1953    0.000021
1954    0.000042
1955    0.000042
1956    0.000104
1957    0.000042
1958    0.000083
1959    0.000146
1960    0.000708
1961    0.000125
1962    0.000083
1963    0.000187
1964    0.000250
1965    0.000354
1966    0.000458
1967    0.000562
1968    0.000541
          ...   
1987    0.001562
1988    0.002957
1989    0.003769
1990    0.008225
1991    0.007413
1992    0.008142
1993    0.009266
1994    0.013743
1995    0.027320
1996    0.030069
1997    0.042230
1998    0.051080
1999    0.062449
2000    0.069842
2001    0.056265
2002    0.052746
2003    0.056785
2004    0.056994
2005    0.062782
2006    0.056390
2007    0.047977
2008    0.046457
2009    0.043687
2010    0.033255
2011    0.034025
2012    0.027549
2013    0.016784
2014    0.0138

Most of the car in the dataset was registered after the 90s.

In [152]:
autos["brand"].value_counts(normalize=True, dropna=False)

volkswagen        0.212128
bmw               0.110031
opel              0.108157
mercedes_benz     0.095350
audi              0.086396
ford              0.069779
renault           0.047352
peugeot           0.029528
fiat              0.025863
seat              0.018179
skoda             0.016034
mazda             0.015139
nissan            0.015097
citroen           0.013910
smart             0.013910
toyota            0.012473
sonstige_autos    0.010953
hyundai           0.009849
volvo             0.009246
mini              0.008642
mitsubishi        0.008142
honda             0.007850
kia               0.007101
alfa_romeo        0.006622
porsche           0.006101
suzuki            0.005914
chevrolet         0.005706
chrysler          0.003665
dacia             0.002561
daihatsu          0.002561
jeep              0.002249
subaru            0.002186
land_rover        0.002041
saab              0.001603
jaguar            0.001583
trabant           0.001562
daewoo            0.001499
r

In [206]:
autos["brand"].value_counts(normalize=True, dropna=False).index

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'mazda', 'nissan', 'citroen',
       'smart', 'toyota', 'sonstige_autos', 'hyundai', 'volvo', 'mini',
       'mitsubishi', 'honda', 'kia', 'alfa_romeo', 'porsche', 'suzuki',
       'chevrolet', 'chrysler', 'dacia', 'daihatsu', 'jeep', 'subaru',
       'land_rover', 'saab', 'jaguar', 'trabant', 'daewoo', 'rover', 'lancia',
       'lada'],
      dtype='object')

The top 3 brands cover more than 40% of the total cars. Therefore, we will aggregate based on the top three brands.

In [204]:
brand_price = {}

for index, row in autos.iterrows():
    b = row["brand"]
    if b in ["volkswagen", "bmw", "opel", "mercedes_benz", "audi", "ford"]: 
        if b in brand_price:
            brand_price[b] += row["price"]
        else:
            brand_price[b] = row["price"]
    else:
        if "others" in brand_price:
            brand_price["others"] += row["price"]
        else:
            brand_price["others"] = row["price"]
        
print(brand_price)

{'others': 94391879.0, 'bmw': 44040265.0, 'volkswagen': 55278559.0, 'ford': 13234518.0, 'mercedes_benz': 38853912.0, 'audi': 37729554.0, 'opel': 14941665.0}


In [215]:
brand_count = {}

for index, row in autos.iterrows():
    b = row["brand"]
    if b in ["volkswagen", "bmw", "opel", "mercedes_benz", "audi", "ford"]: 
        if b in brand_count:
            brand_count[b] += 1
        else:
            brand_count[b] = 1
    else:
        if "others" in brand_count:
            brand_count["others"] += 1
        else:
            brand_count["others"] = 1

In [228]:
brand_avg_price = {}
for brand, price in brand_price.items():
    for k, v in brand_count.items():
        if brand == k:
            brand_avg_price[brand]=price/v
brand_sort = []
for brand, avg_price in brand_avg_price.items():
    brand_sort.append([avg_price, brand])
brand_sort.sort(reverse=True)

for item in brand_sort:
    print(f'Average price of {item[1].upper()}: {item[0]:.2f} US$')


Average price of AUDI: 9093.65 US$
Average price of MERCEDES_BENZ: 8485.24 US$
Average price of BMW: 8334.65 US$
Average price of OTHERS: 6177.88 US$
Average price of VOLKSWAGEN: 5426.38 US$
Average price of FORD: 3949.42 US$
Average price of OPEL: 2876.72 US$
